[AC-447] Add Data Validation when registering patients

f4ww4z opened the pull request:
https://issues.openmrs.org/browse/AC-447

Added a function in `ViewUtils` to validate Strings to check for illegal characters in text fields. Also added tests to the presenter.

Screenshots of the UI are [name_errors](https://issues.openmrs.org/secure/attachment/51120/name_errors.png) and [invalid_address](https://issues.openmrs.org/secure/attachment/51119/address_error.png).

shivtej1505 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java:
```diff
@@ -240,6 +257,36 @@ private Person createPerson() {
         person.setAddresses(addresses);
 
+        // Validate names
+        String unvalidatedError = getString(R.string.name_invalid_error);
```
Give another name

shivtej1505 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java:
```diff
@@ -240,6 +257,36 @@ private Person createPerson() {
+        // Validate names
+        String unvalidatedError = getString(R.string.name_invalid_error);
+        String middleNameError = getString(R.string.midname_invalid_error);
+        String givenMiddleNameError = getString(R.string.fmname_invalid_error);
```
Can we clearly mention in which field contain invalid characters? Like here, we didn't include anything about `given name` and `family name` so that might be confusing for users.

shivtej1505 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java:
```diff
@@ -240,6 +257,36 @@ private Person createPerson() {
+        if (ViewUtils.isEmpty(edfname)) {
+            // Require first name
+            fnameerror.setText(emptyError);
+        } else if (!ViewUtils.validateText(ViewUtils.getInput(edfname), ViewUtils.ILLEGAL_CHARACTERS)) {
```
We can put this line in separate function as it is being used in at least 3 places.

shivtej1505 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java:
```diff
@@ -240,6 +257,36 @@ private Person createPerson() {
+        if (!ViewUtils.isEmpty(edmname) &&
+                !ViewUtils.validateText(ViewUtils.getInput(edmname), ViewUtils.ILLEGAL_CHARACTERS)) {
+            // SHow invalid entry for both text
+            if (fnameerror.getText().toString().equals(unvalidatedError)) {
```
We should omit this kind of logic, as this statement check the string which is there because of invalid `first name`. We should directly check if the `first name` field is invalid or not, we shouldn't check the consequence due to the invalid field.

shivtej1505 replied:
Also, you can first check for `middle name` and `first name` field together. It would be more readable and understandable logic.

shivtej1505 commented on openmrs-client/src/main/java/org/openmrs/mobile/utilities/ViewUtils.java:
```diff
@@ -16,8 +16,14 @@
+
 public class ViewUtils {
 
+    public static final String ILLEGAL_CHARACTERS = "[$&+,:;=\\\\?@#|/'<>.^*()%!-]";
```
What about the names like `Jean-paul Sartre` and `John Q. Public`.
See here: https://www.google.co.in/search?q=jean-paul+sartre

f4ww4z replied:
I just saw https://www.w3.org/International/questions/qa-personal-names and I think `,` and `-` should be allowed. @shivtej1505 Please tell me if I am missing characters which are also allowed in names.

f4ww4z commented:
@shivtej1505 I have made your required changes. I removed `,`, `.` and `-` from `ILLEGAL_CHARACTERS` and removed the `.` from `ADDRESS`. Also made the if statements easier to read.
